bgpd/bgp_labelpool.c

```diff
@@ -0,0 +1,539 @@
+}
+
+void lp_init(struct thread_master *master, struct labelpool *pool)
+{
+	if (BGP_DEBUG(vpn, VPN_LEAK_LABEL))
+		zlog_debug("%s: entry", __func__);
+
+	lp = pool;	/* Set module pointer to pool data */
+
+	lp->ledger = skiplist_new(0, NULL, lp_lcb_free);
+	lp->inuse = skiplist_new(0, NULL, NULL);
+	lp->chunks = list_new();
+	lp->chunks->del = lp_chunk_free;
+	LABEL_FIFO_INIT(&lp->requests);
+	lp->callback_q = work_queue_new(master, "label callbacks");
+	if (!lp->callback_q) {
+		zlog_err("%s: Failed to allocate work queue", __func__);
+		exit(1);
+	}
+
+	lp->callback_q->spec.workfunc = lp_cbq_docallback;
+	lp->callback_q->spec.del_item_data = lp_cbq_item_free;
+	lp->callback_q->spec.max_retries = 0;
+}
+
+void lp_finish(void)
+{
+	struct lp_fifo *lf;
+
+	if (!lp)
+		return;
+
+	skiplist_free(lp->ledger);
+	lp->ledger = NULL;
+
+	skiplist_free(lp->inuse);
+	lp->inuse = NULL;
+
+	list_delete_and_null(&lp->chunks);
+
+	while ((lf = LABEL_FIFO_HEAD(&lp->requests))) {
+
+		LABEL_FIFO_DEL(&lp->requests, lf);
+		XFREE(MTYPE_BGP_LABEL_FIFO, lf);
+	}
+
+	work_queue_free_and_null(&lp->callback_q);
+
+	lp = NULL;
+}
+
+static mpls_label_t get_label_from_pool(void *labelid)
+{
+	struct listnode *node;
+	struct lp_chunk *chunk;
+	int debug = BGP_DEBUG(vpn, VPN_LEAK_LABEL);
+
+	/*
+	 * Find a free label
+	 * Linear search is not efficient but should be executed infrequently.
+	 */
+	for (ALL_LIST_ELEMENTS_RO(lp->chunks, node, chunk)) {
+		uintptr_t lbl;
+
+		if (debug)
+			zlog_debug("%s: chunk first=%u last=%u",
+				__func__, chunk->first, chunk->last);
+
+		for (lbl = chunk->first; lbl <= chunk->last; ++lbl) {
+			/* labelid is key to all-request "ledger" list */
+			if (!skiplist_insert(lp->inuse, (void *)lbl, labelid)) {
+				/*
+				 * Success
+				 */
+				return lbl;
+			}
+		}
+	}
+	return MPLS_LABEL_NONE;
+}
+
+/*
+ * Success indicated by value of "label" field in returned LCB
+ */
+static struct lp_lcb *lcb_alloc(
+	int	type,
+	void	*labelid,
+	int	(*cbfunc)(mpls_label_t label, void *labelid, int lost))
+{
+	/*
+	 * Set up label control block
+	 */
+	struct lp_lcb *new = XCALLOC(MTYPE_BGP_LABEL_CB,
+		sizeof(struct lp_lcb));
+
+	assert(new);
+	new->label = get_label_from_pool(labelid);
+	new->type = type;
+	new->labelid = labelid;
+	new->cbfunc = cbfunc;
+
+	return new;
+}
+
+/*
+ * Callers who need labels must supply a type, labelid, and callback.
+ * The type is a value defined in bgp_labelpool.h (add types as needed).
+ * The callback is for asynchronous notification of label allocation.
+ * The labelid is passed as an argument to the callback. It should be unique
+ * to the requested label instance.
+ *
+ * If zebra is not connected, callbacks with labels will be delayed
+ * until connection is established. If zebra connection is lost after
+ * labels have been assigned, existing assignments via this labelpool
+ * module will continue until reconnection.
+ *
+ * When connection to zebra is reestablished, previous label assignments
```

Comment:
This comment seems strange to me.  Currently the only real model for loosing zebra connection is if zebra goes down.  When this happens daemons do not reconnect to zebra as that we do not have a proper recovery mechanism for the routes that are lost currently.  In fact we typically restart all the daemons when this happens. 

What is your experience here that leads you to say the zebra connection is re-established?

Reply:
Primary motivation was to support either order of startup bgpd vs. zebra, as well as to prevent blocking bgpd in cases of no bgpd at all. Once that is handled, supporting loss/reconnect of zebra is a small step.

Reply:
My recollection about this is that the client daemons do reconnect to zebra when it restarts, but this is not very well tested (and not expected to happen) so lots of inconsistencies can happen. My understanding is that zebra is a single point of failure and it should be solid as a rock for this reason.